core/node/api_server/src/web3/namespaces/en.rs

```diff
@@ -14,16 +15,60 @@ use crate::web3::{backend_jsonrpsee::MethodTracer, state::RpcState};
 #[derive(Debug)]
 pub(crate) struct EnNamespace {
     state: RpcState,
+    /// First batch to commit to L1 simulated by the main node.
+    /// This is temporary and used only for testing L1 batch signing by consensus attesters.
+    first_batch_to_commit: L1BatchNumber,
+}
+
+fn to_l1_batch_number(n: attester::BatchNumber) -> anyhow::Result<L1BatchNumber> {
+    Ok(L1BatchNumber(
+        n.0.try_into().context("L1BatchNumber overflow")?,
+    ))
 }
 
 impl EnNamespace {
-    pub fn new(state: RpcState) -> Self {
-        Self { state }
+    pub async fn new(state: RpcState) -> anyhow::Result<Self> {
+        let first_batch_to_commit = async {
+            let mut conn = state.acquire_connection().await.context("connection()")?;
+            // Try to continue from where we left.
+            if let Some(last) = conn
+                .consensus_dal()
+                .get_last_batch_certificate_number()
+                .await
+                .context("get_last_batch_certificate_number()")?
+            {
+                return to_l1_batch_number(last + 1);
+            }
+            // Otherwise start with the next sealed L1 batch.
+            if let Some(sealed) = conn
+                .blocks_dal()
+                .get_sealed_l1_batch_number()
+                .await
+                .context("get_sealed_l1_batch_number()")?
+            {
+                return Ok(sealed + 1);
+            }
+            // Otherwise start from the first non-pruned batch.
+            let info = conn
+                .pruning_dal()
+                .get_pruning_info()
+                .await
+                .context("get_pruning_info()")?;
+            Ok(info
+                .last_soft_pruned_l1_batch
+                .map(|n| n + 1)
+                .unwrap_or(L1BatchNumber(0)))
+        }
```

Comment:
Note that the consensus starting point will not be the same as the first stored batch otherwise we would need to backfill information which doesn't make sense. This by itself requires inspecting l1_batches and l1_batches_certificates. pruning_info encapsulates everything else, which IMO is good enough abstraction.

Reply:
It would definitely be nice to move this to the DAL because I don't think I would get it right. I got as far as looking at the [snapshot_recovery_dal](https://github.com/matter-labs/zksync-era/blob/f1cbb74b863b6e0bcfa74ad780beef29844bac6e/core/node/consensus/src/storage/connection.rs#L394) but didn't know about the `pruning_dal`. A bit too much for the `web3` IMO.

Reply:
I've removed the pruning_dal lookup

Reply:
And snapshots don't have to be considered because the main node doesn't have them?

Reply:
yes

Reply:
The pruning_info lookup accounted for the snapshots on the main node (not pruning of the main node), but since there are no snapshots on the main node, I've removed the corresponding test case as well.